dvc/external_repo.py

```diff
@@ -65,66 +68,211 @@ def clean_repos():
+        fetch_infos = [self._fetch_info(path_info)]
+        if self.repo_tree.isdir(path_info) and not self.repo_tree.isdvc(
+            path_info
+        ):
+            fetch_infos.extend(self._recursive_outputs(path_info, recursive))
 
-            fs_copy(fspath(path_info), fspath(to_info))
-        except FileNotFoundError:
-            raise PathMissingError(path, self.url)
+        self._fetch_external(fetch_infos, **kwargs)
 
-    def _pull_cached(self, out, path_info, dest):
         with self.state:
-            tmp = PathInfo(tmp_fname(dest))
-            src = tmp / path_info.relative_to(out.path_info)
+            self.repo_tree.copytree(path_info, to_info)
 
-            out.path_info = tmp
+    def fetch_external(self, files, **kwargs):
+        """Fetch erepo files into the specified cache.
 
-            # Only pull unless all needed cache is present
-            if out.changed_cache(filter_info=src):
-                self.cloud.pull(out.get_used_cache(filter_info=src))
+        Works with files tracked by Git and DVC.
+        """
+        files = [(PathInfo(self.root_dir) / name, None) for name in files]
+        return self._fetch_external(files, **kwargs)
+
+    def _fetch_external(self, fetch_infos, **kwargs):
+        downloaded, failed = 0, 0
 
+        with self.state:
+            for path_info, filter_info in fetch_infos:
+                if self.repo_tree.isdvc(path_info):
+                    (out,) = self.find_outs_by_path(path_info, strict=False)
+                    d, f = self._fetch_out(
+                        out, filter_info=filter_info, **kwargs
+                    )
+                else:
+                    d, f = self._fetch_git(path_info)
+                downloaded += d
+                failed += f
+
+        if failed:
+            logger.exception(
+                "failed to fetch '{}' files from '{}' repo".format(
+                    failed, self.url
+                )
+            )
+        return downloaded, failed
+
+    def _fetch_out(self, out, filter_info=None, **kwargs):
+        """Fetch specified erepo out."""
+        downloaded, failed = 0, 0
+        if out.changed_cache(filter_info=filter_info):
+            used_cache = out.get_used_cache()
             try:
-                out.checkout(filter_info=src)
-            except CheckoutError:
-                raise FileNotFoundError
+                downloaded += self.cloud.pull(used_cache, **kwargs)
+            except DownloadError as exc:
+                failed += exc.amount
+        return downloaded, failed
+
+    def _fetch_git(self, path_info):
+        """Copy git tracked file into specified cache."""
+        downloaded, failed = 0, 0
+        if hasattr(self, "cache"):
+            local_cache = self.cache.local
+        elif self._local_cache:
+            local_cache = self._local_cache
+        else:
+            return downloaded, failed
+
+        info = local_cache.save_info(path_info)
```

Comment:
Looks like we still depend on checked-out git repo, right? Because path_info needs to point to an existing git file. I though we could use `tree` directly to copy obj directly from index to our cache.

Reply:
`path_info` points to a path that only exists in the git tree, so `remote.save_info` will return a checksum computed from the git-index obj. Then we do
```
elif local_cache(changed_cache...):
    local_cache.save_tree(...)
```
which copy's the obj directly from git index into our cache (if it hasn't already been copied/fetched)

Reply:
Oh, got it, I didn't catch that initially.